patches/api/0419-Player-standing-on-position-API.patch

```diff
@@ -0,0 +1,55 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
+Date: Fri, 9 Jun 2023 22:49:45 -0400
+Subject: [PATCH] Player standing on position API
+
+
+diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
+index a2a423d4e4c2702ba5967223cab0432dd7d04732..6fa7d251f54be9d0ef6b0ec0283efa927bae1a0b 100644
+--- a/src/main/java/org/bukkit/entity/Entity.java
++++ b/src/main/java/org/bukkit/entity/Entity.java
+@@ -954,4 +954,44 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
+      */
+     boolean wouldCollideUsing(@NotNull BoundingBox boundingBox);
+     // Paper End - Collision API
++    // Paper start - Player standing on position API
++    /**
++     * Gets the block that is currently being used to calculate movement affects (i.e friction) on the entity.
++     * <p>
++     * Examples:
++     * - When sanding on a slab under an ice block, this will return the position of the ice block.
++     * - When flying on top of a flower located under a grass block this will return the position of the grass block (although this technically will not affect your movement).
++     *
++     * @return block position used for movement affects
++     */
++    @NotNull
++    io.papermc.paper.math.BlockPosition getBlockPosBelowThatAffectsMyMovement();
++
++    /**
++     * Gets the block that is currently under the entity.
++     * This will return {@link Entity#getSupportingPos()} as long as it is not null,
++     * else it will instead return the bottom most block at the entity's hitbox.
++     * <p>
++     * Examples:
++     * - When sanding on a slab under an ice block, this will return the position of the slab.
++     * - When flying on top of a flower located under a grass block this will return the position of the corn flower.
++     *
++     * @return {@link Entity#getSupportingPos()} if not null or the entity's bottom most block
++     */
++    @NotNull
++    io.papermc.paper.math.BlockPosition getOnPos();
```

Comment:
I think this would probably be more useful if it returned a `Block`. The javadocs even say `Gets the block...`. I think the usual usecase is gonna be to check the type. CraftBlock is just a wrapper so cheap to create.

Reply:
Not exactly because it doesn't always represent a block actually. See the javadoc comments, where depending on the entity's position it returns the bottom of their hitbox.

Reply:
I am not sure actually, opinions?

Reply:
These values always apply to the world the player is in correct? If that is the case, Block or Location seems more appropriate. Giving extra context in a returned value is fine. The problem with Location usually shows up when it is a parameter rather than a return value.

Reply:
Hmm okay, I think Location is fine then. I would just rather stay to more primitive types.